src/app/modules/group/components/associated-item/associated-item.component.ts

```diff
@@ -0,0 +1,138 @@
+  isNewAssociatedItem,
+  getAllowedTypesForNewAssociatedItem,
+} from './associated-item-types';
+import { errorIsHTTPForbidden } from 'src/app/shared/helpers/errors';
+import { mapToFetchState } from 'src/app/shared/operators/state';
+
+@Component({
+  selector: 'alg-associated-item',
+  templateUrl: './associated-item.component.html',
+  styleUrls: [ './associated-item.component.scss' ],
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: forwardRef(() => AssociatedItemComponent),
+      multi: true,
+    }
+  ]
+})
+export class AssociatedItemComponent implements ControlValueAccessor, OnChanges, OnDestroy {
+  @Input() contentType: ItemTypeCategory = 'activity';
+
+  allowedNewItemTypes: NewContentType<ItemType>[] = [];
+
+  private readonly contentType$ = new ReplaySubject<ItemTypeCategory>();
+
+  private readonly itemChanges$ = new ReplaySubject<{
+    item: NoAssociatedItem|NewAssociatedItem|(ExistingAssociatedItem&{ name?: string }),
+    triggerChange: boolean,
+  }>();
+
+  private refresh$ = new Subject<void>();
+  readonly state$ = combineLatest([
+    this.itemChanges$,
+    this.contentType$,
+  ]).pipe(
+    distinctUntilChanged(),
+    switchMap(([ data, contentType ]) => {
+      if (data.triggerChange) this.onChange(data.item);
+
+      if (!isExistingAssociatedItem(data.item)) {
+        return of({
+          tag: data.item.tag, id: undefined, path: null,
+          name: isNewAssociatedItem(data.item) ? data.item.name : undefined
+        });
+      }
+
+      const id = data.item.id;
+      const name = data.item.name !== undefined ? of(data.item.name) :
+        this.getItemByIdService.get(id).pipe(map(item => item.string.title));
+
+      return name.pipe(
+        map(name => ({
+          tag: 'existing-item',
+          id,
+          name,
+          path: urlArrayForItemRoute(rawItemRoute(contentType, id))
+        })),
+        catchError(err => {
+          if (errorIsHTTPForbidden(err)) return of({
+            tag: 'existing-item',
+            name: $localize`You don't have access to this ` + contentType === 'activity' ? $localize`activity` : $localize`skill` + '.',
+            path: null,
+          });
+          throw err;
+        })
+      );
+    }),
+    mapToFetchState({ resetter: this.refresh$ }),
+  );
+
+  private onChange: (value: NoAssociatedItem|NewAssociatedItem|ExistingAssociatedItem) => void = () => {};
+
+  searchFunction = (value: string): Observable<AddedContent<ItemType>[]> =>
+    this.contentType$.pipe(
+      switchMap(contentType =>
+        this.searchItemService.search(value, contentType === 'activity' ? [ 'Chapter', 'Course', 'Task' ] : [ 'Skill' ])
+      )
+    );
+
+  constructor(
+    private getItemByIdService: GetItemByIdService,
+    private searchItemService: SearchItemService,
+  ) { }
+
+  ngOnChanges(): void {
+    this.allowedNewItemTypes = getAllowedTypesForNewAssociatedItem(this.contentType);
+    this.contentType$.next(this.contentType);
```

Comment:
There is no reason for contentType to change while it is in use...  I think using a subject can introduce (uselessly) new behaviors ( for instance your search function do not complete anymore) and so new issues (especially if you really change contentType... what you probably haven't tested)...

So I would rather:
- force `contentType` to be set for the component
- if it changes, throw an error
- use just a var, not a subject

Reply:
There are places in stream where I still need content type. So I've add only error handling, but it looks very unexpected case as content type added in template is just string.

Pushing content type event in ngOnChanges instead onInit - is just because it goes first in lifehook

Reply:
Or do you mean I can use `this.contentType` directly in `switchMap` section?

Reply:
Fixed